### PR TITLE
Create e2e test report issue template

### PR DESCRIPTION
**Pull request by cesarcosta99**
#### Changes proposed in this Pull Request

This PR adds a new issue template to be used when logging broken e2e tests. This is part of a new process to log issues in e2e tests to reduce disruption from flaky or tricky tests, and over time better understand how our e2e tests are failing.

The process consists in:

1. Logging of persistent failures (this issue template is going to facilitate this step).
2. Assignment of ownership: issue will be triaged and assigned to the relevant team.
3. Resolution.

More context in paJDYF-cCK-p2.

#### Testing instructions

Not applicable.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8420` or branch name `chore/e2e-test-report-template` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=chore/e2e-test-report-template) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a11852c0651923bca37ebf7b9387e18a44187d72
- Build time: 2024-03-21 17:07:34 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: .github/ISSUE_TEMPLATE/e2e_test_report.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+<!-- Add a clear and concise description of what the issue is. Please, be as descriptive as possible. -->
+
+**Error**:
+<!-- The error that the e2e test indicates. -->
+
+**Spec**:
+<!-- The test file that's triggering this error. -->
+
+**Relevant line(s) of code**:
+<!-- Line of code that is triggering this error. -->
```
Comment (haszari) [inline .github/ISSUE_TEMPLATE/e2e_test_report.md]: Is this trivial? I like that there are prompts here to capture the relevant information. but at the same time it's important to make this a zero-minute frictionless process.

If the test failures are visible in GitHub actions logs, maybe it's extra bureaucracy to require reporter to extract specific details and paste here?

TLDR I think the prompts are useful but we should be lenient on what we expect. 

Some options:
- Ask the reporter to paste the relevant chunk of the output from the test failure. This means we expect them to at least look at the logs, and paste in a chunk that covers the prompts above.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+**GitHub jobs where this spec is failing**:
+<!-- Can you list all the jobs where the test failing? e.g. "WC - latest | wcpay - shopper", "WC - beta | wcpay - shopper" etc. -->
```
Comment (haszari) [inline .github/ISSUE_TEMPLATE/e2e_test_report.md]: This is a bit manual, this would slow me down if I was reporting. Is there a way we can paste this in? 

Worst case (I don't really like this) we could ask the reporter to paste a screenshot of the action checklist.

Another option would be to ask them to mention the first job that fails (i.e. name at least one), and the number of jobs that are failing/affected.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+- 
+
+### Additional context
+<!-- Any additional context or details you think might be helpful. -->
+<!-- Ticket numbers/links, plugin versions, system statuses etc. -->
+
+---
+
+- [ ] Added priority label.
+<!-- For the priority label, use your best judgement. -->
```
Comment (haszari) [inline .github/ISSUE_TEMPLATE/e2e_test_report.md]: This a good prompt, I think in many cases the reporter will have a reasonable idea of priority. Possibly even from the test name (i.e. if it's a critical flow or a secondary flow, etc). 

Even better would be to add this as a prompt so reporter can explain their reasoning. For example:

```
### Priority
<!-- add a priority label based on your best judgement (quick is fine) -->
<!-- optional: add comments here to explain your decision or highlight any questions/potential risks -->
````

File: .github/ISSUE_TEMPLATE/e2e_test_report.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+- [ ] Added e2e label.
+<!-- Add one of the "e2e: broken flow/test/environment" labels. See their description and try to match it with the actual issue. -->
```
Comment (haszari) [inline .github/ISSUE_TEMPLATE/e2e_test_report.md]: I don't think we should expect this – it makes this a ~30 min task rather than a ~10 min task. I think it should be optional, and required when the issue is closed. 

I understand our concern here is that we want accurate data. Worst case: if it's missing (😮), it should be simple for e.g. the person who wants accurate data to add it later, by reading the fix PR or possibly chatting to someone (i.e. if their fix PR was not documented clearly). TLDR an audit process (later) would work better here than more expectations at the start (on reporter).

Let's outline the whole process required of a broken e2e reporter, and be realistic about how that could disrupt their flow. My dream is that it's zero friction, in reality we have a few tasks, each with multiple steps:

1. Investigate to work out which test(s) need skipping.
2. Push a PR to skip the broken test.
3. Open an issue to log that the test is broken and needs follow up (i.e. use this template).

My expectation is that this would take 5-10 min in the best possible case, and 15-30 would be a good allowance. Requiring an accurate or even best-guess `e2e-fail` label adds overhead.

Side note – maybe we should use `e2e-fail` or `e2e-broken` prefix for these new labels.

**Review by haszari — APPROVED** [review]
Thanks for setting this up @cesarcosta99 ! I've added a few comments with my thoughts.

The most important thing from my point of view is making this a low-friction process so e2e reporters can get back their focus (e.g. project work). 

Pre-approving, I am happy for you to own this and incorporate my feedback alongside others 🙌🏻

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 1 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: Good call out! I've adopted your suggestion and changed it in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/e6d7a0276b8b9e11c6bd7f7d2d94e12f0a9d2108.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 2 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: Actually, this is not as easy as it sounds. At first, I thought about implementing a list of checkboxes so the reporter could just mark the ones that match the failing jobs. However, we have 16 items and this isn't practical. Furthermore, the jobs may change with time, making this list obsolete.

I don't like the screenshot option because there might be jobs failing in the middle that are failing for different reasons.

I ended up removing this in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/7b0a659931632efa1ffe560d13994b7ec4452a49 since the assignee will see what jobs are failing while fixing it and check that information anyway.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 3 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: Applied this in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/aa8441ac331c55109d9354eaa70708f05b510598.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 4 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: I still think it's good to have an initial input by the logger, or we may end up in the worst case, where the label is missing.

>it should be simple for e.g. the person who wants accurate data to add it later, by reading the fix PR or possibly chatting to someone

I don't think this is an easy undertake, especially for someone who doesn't have context on the issue nor on the fix. Imagine a scenario where this person have dozens of issues like this and none of them have this label? They'd take a couple of days to gather all that information.

I sort of mixed your previous comment idea with this one and came into a middle-ground option. That would be asking the logger to label it using a quick judgement, so we don't take much of their time. At the same time, we leave an explicit ask for the assignee, who will have most context and more accurate information, to review this input.

If the assignee mark the checkbox, no further action is needed and we have accurate data. Otherwise, everyone's aware that this label may not be accurate, and anyone can ping the assignee later.

I applied this change in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/aa8441ac331c55109d9354eaa70708f05b510598 and it will render like this:

image

Comment (haszari) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: Sounds good. Adding the label should get easier as it gets more close to fixed.

Comment (haszari) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: That checkbox wording is a little strange - because it assumes the label is usually correct. I would just ask what we want – say something like:

- Please ensure when closing this issue (PR fix) that only one `e2e-fail` label is added and it is accurate. 

Keep it simple.

There's no bulletproof way to get these labels right with wording or process, ultimately we have to ask and trust (and be prepared to fix errors).

File: changelog/chore-e2e-test-report-template [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Significance: patch
+Type: dev
+Comment: Create e2e test report issue template.
```
Comment (vbelolapotkov) [inline changelog/chore-e2e-test-report-template:3]: I think we don't need this changelog entry to be pulled into release notes (changelog). IIRC there is a recommended way of skipping the changelog entry

File: .github/ISSUE_TEMPLATE/e2e_test_report.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+about: Report an e2e test failure.
+title: ''
+labels: ['category: e2e', 'needs triage']
+assignees: ''
+
+---
+
+### Description
+<!-- Before reporting an issue, are you sure this e2e test is failing consistently? -->
+<!-- Add a clear and concise description of what the issue is. Please, be as descriptive as possible. -->
```
Comment (vbelolapotkov) [inline .github/ISSUE_TEMPLATE/e2e_test_report.md:12]: Let's explicitly ask to list tests/specs which are failing and link to the job when it started failing consistently.

**Review by vbelolapotkov — APPROVED** [review]
Thanks for working on this @cesarcosta99 . I left some comments, but most of the things look good. 

I'm approving it, because it looks good enough for the first version. We'll be able to iterate on it if there are any major gaps left.

File: changelog/chore-e2e-test-report-template (hunk 5 above)
Comment (cesarcosta99) [reply changelog/chore-e2e-test-report-template:3]: This is the recommended way, it won't be pulled into the changelog 👍

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 6 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md:12]: Added it in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/3e2cfd674d2baab256a6cdaf2a76a50ef2d31235.

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 4 above)
Comment (cesarcosta99) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md]: I adjusted it to your suggestion in https://github.com/Automattic/woocommerce-payments/pull/8420/commits/e542a5cc2ca7e2e779f8831e93454359b3e66318. I agree there's no bulletproof method, but I left the checkbox because I think it will help the assignee to visualize that todo and, when analyzing data, we can be sure the label has been reviewed.

File: changelog/chore-e2e-test-report-template (hunk 5 above)
Comment (vbelolapotkov) [reply changelog/chore-e2e-test-report-template:3]: TIL, thanks!

File: .github/ISSUE_TEMPLATE/e2e_test_report.md (hunk 6 above)
Comment (vbelolapotkov) [reply .github/ISSUE_TEMPLATE/e2e_test_report.md:12]: Looks good, thanks!
